text/0000-variadic.md

```diff
@@ -0,0 +1,287 @@
+code, extracting arguments from that structure requires exceptionally
+error-prone, platform-specific code, for which the crates.io ecosystem provides
+only partial solutions for a few target architectures.
+
+This RFC does not propose an interface intended for native Rust code to pass
+variable numbers of arguments to a native Rust function, nor an interface that
+provides any kind of type safety. This proposal exists primarily to allow Rust
+to provide interfaces callable from C code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+C code allows declaring a function callable with a variable number of
+arguments, using an ellipsis (`...`) at the end of the argument list. For
+compatibility, unsafe Rust code may export a function compatible with this
+mechanism.
+
+Such a declaration looks like this:
+
+```rust
+pub unsafe extern "C" fn func(arg: T, arg2: T2, ...) {
+    // implementation
+}
+```
+
+The `...` at the end of the argument list declares the function as variadic.
+The function must use `extern "C"`, and must use `unsafe`. To expose
+such a function as a symbol for C code to call directly, the function may want
+to use `#[no_mangle]` as well; however, Rust code may also pass the function to
+C code expecting a function pointer to a variadic function.
+
+Unsafe Rust code can also define a variadic closure:
+
+```rust
+let closure = |arg, arg2, ...| {
+    // implementation
+};
+```
+
+Rust code may pass a variadic closure to C code expecting a pointer to a
+variadic function.
+
+To access the arguments, Rust provides the following public interfaces in
+`core::intrinsics` (also available via `std::intrinsics`):
+
+```rust
+/// The argument list of a C-compatible variadic function, corresponding to the
+/// underlying C `va_list`. Opaque.
+pub struct VaList<'a>;
+
+impl<'a> VaList<'a> {
+    /// Obtain the variable arguments of the current function. Produces a
+    /// compile-time error if called from a non-variadic function. The compiler
+    /// will supply the appropriate lifetime when called, and prevent that
+    /// lifetime from outliving the variadic function.
+    pub unsafe fn start() -> VaList<'a>;
+
+    /// Extract the next argument from the argument list.
+    pub unsafe fn<T: VaArg> arg(&mut self) -> T;
+}
+
+impl<'a> Clone for VaList<'a>;
+impl<'a> Drop for VaList<'a>;
+
+/// The type of arguments extractable from VaList
+trait VaArg;
+
+impl VaArg for i8;
+impl VaArg for i16;
+impl VaArg for i32;
+impl VaArg for i64;
+impl VaArg for isize;
+
+impl VaArg for u8;
+impl VaArg for u16;
+impl VaArg for u32;
+impl VaArg for u64;
+impl VaArg for usize;
+
+impl VaArg for f32;
+impl VaArg for f64;
+
+impl<T> VaArg for *const T;
+impl<T> VaArg for *mut T;
+```
+
+The `libc` crate additionally provides implementations of the `VaArg` trait for
+the raw C types corresponding to the Rust integer and float types above:
+
+```rust
+impl VaArg for c_char;
```

Comment:
Unless these type aliases became concrete types, I don't think the `std` and `libc` crates need to provide any additional impls.

Reply:
Fixed.